File: frontend/src/lib/subscription/feature-registry.ts
// Feature types and definitions matching the backend
export type FeatureType = 'limit' | 'flag' | 'custom';

export interface FeatureDefinition {
  key: string;
  type: FeatureType;
  displayName: string;
  description?: string;
  unit?: string;
  unlimitedText?: string;
  format?: string;
  icon?: string;
  category?: string;
  sortOrder: number;
}

// Feature keys matching backend constants
export const LIMITS = {
  RESTAURANTS: 'max_restaurants',
  QR_CODES: 'max_qr_codes',
  FEEDBACKS_PER_MONTH: 'max_feedbacks_per_month',
  TEAM_MEMBERS: 'max_team_members'
} as const;

export const FLAGS = {
  BASIC_ANALYTICS: 'basic_analytics',
  ADVANCED_ANALYTICS: 'advanced_analytics',
  FEEDBACK_EXPLORER: 'feedback_explorer',
  CUSTOM_BRANDING: 'custom_branding',
  PRIORITY_SUPPORT: 'priority_support'
} as const;

// Feature registry with all feature definitions
export const featureRegistry: Record<string, FeatureDefinition> = {
  // Limits
  [LIMITS.RESTAURANTS]: {
    key: LIMITS.RESTAURANTS,
    type: 'limit',
    displayName: 'Restaurants',
    description: 'Maximum number of restaurants',
    unit: 'restaurants',
    unlimitedText: 'Unlimited restaurants',
    format: '{value} restaurant(s)',
    icon: 'store',
    category: 'core',
    sortOrder: 1
  },
  [LIMITS.QR_CODES]: {
    key: LIMITS.QR_CODES,
    type: 'limit',
    displayName: 'QR Codes',
    description: 'Total QR codes across all restaurants',
    unit: 'QR codes',
    unlimitedText: 'Unlimited QR codes',
    format: '{value} QR codes',
    icon: 'qr-code',
    category: 'core',
    sortOrder: 2
  },
  [LIMITS.FEEDBACKS_PER_MONTH]: {
    key: LIMITS.FEEDBACKS_PER_MONTH,
    type: 'limit',
    displayName: 'Monthly Feedbacks',
    description: 'Maximum feedbacks per month',
    unit: 'feedbacks/month',
    unlimitedText: 'Unlimited feedbacks',
    format: '{value} feedbacks/month',
    icon: 'message-square',
    category: 'core',
    sortOrder: 3
  },
  [LIMITS.TEAM_MEMBERS]: {
    key: LIMITS.TEAM_MEMBERS,
    type: 'limit',
    displayName: 'Team Members',
    description: 'Maximum team members',
    unit: 'members',
    unlimitedText: 'Unlimited team members',
    format: '{value} team member(s)',
    icon: 'users',
    category: 'collaboration',
    sortOrder: 4
  },

  // Flags
  [FLAGS.BASIC_ANALYTICS]: {
    key: FLAGS.BASIC_ANALYTICS,
    type: 'flag',
    displayName: 'Basic Analytics',
    description: 'View feedback analytics and insights',
    icon: 'bar-chart-2',
    category: 'analytics',
    sortOrder: 20
  },
  [FLAGS.ADVANCED_ANALYTICS]: {
    key: FLAGS.ADVANCED_ANALYTICS,
    type: 'flag',
    displayName: 'Advanced Analytics',
    description: 'Detailed insights and reporting',
    icon: 'bar-chart',
    category: 'analytics',
    sortOrder: 21
  },
  [FLAGS.FEEDBACK_EXPLORER]: {
    key: FLAGS.FEEDBACK_EXPLORER,
    type: 'flag',
    displayName: 'Feedback Explorer',
    description: 'Browse and search all feedback',
    icon: 'search',
    category: 'analytics',
    sortOrder: 22
  },
  [FLAGS.CUSTOM_BRANDING]: {
    key: FLAGS.CUSTOM_BRANDING,
    type: 'flag',
    displayName: 'Custom Branding',
    description: 'Customize with your brand',
    icon: 'palette',
    category: 'customization',
    sortOrder: 23
  },
  [FLAGS.PRIORITY_SUPPORT]: {
    key: FLAGS.PRIORITY_SUPPORT,
    type: 'flag',
    displayName: 'Priority Support',
    description: '24/7 priority customer support',
    icon: 'headphones',
    category: 'support',
    sortOrder: 24
  }
};

// Helper functions
export function getFeatureDefinition(key: string): FeatureDefinition | undefined {
  return featureRegistry[key];
}

export function getFeaturesByCategory(category: string): FeatureDefinition[] {
  return Object.values(featureRegistry)
    .filter(def => def.category === category)
    .sort((a, b) => a.sortOrder - b.sortOrder);
}

export function formatFeatureValue(key: string, value: number | boolean): string {
  const def = featureRegistry[key];
  if (!def) return '';

  if (def.type === 'limit') {
    const limitValue = value as number;
    if (limitValue === -1) {
      return def.unlimitedText || 'Unlimited';
    }
    
    let result = def.format || '{value} {unit}';
    result = result.replace('{value}', limitValue.toLocaleString());
    result = result.replace('{unit}', def.unit || '');
    return result;
  }

  if (def.type === 'flag' && value === true) {
    return def.displayName;
  }

  return '';
}

// Get all features from a plan in display order
export function getPlanFeatures(plan: any): string[] {
  const features: string[] = [];
  
  // Add base features available for all plans
  features.push('Analytics Dashboard');
  features.push('Feedback Explorer');
  
  // Process limits
  if (plan.features?.limits) {
    Object.entries(plan.features.limits)
      .map(([key, value]) => ({
        key,
        value,
        def: featureRegistry[key]
      }))
      .filter(item => item.def)
      .sort((a, b) => a.def.sortOrder - b.def.sortOrder)
      .forEach(item => {
        const formatted = formatFeatureValue(item.key, item.value as number);
        if (formatted) features.push(formatted);
      });
  }
  
  // Process flags
  if (plan.features?.flags) {
    Object.entries(plan.features.flags)
      .map(([key, value]) => ({
        key,
        value,
        def: featureRegistry[key]
      }))
      .filter(item => item.def && item.value === true)
      .sort((a, b) => a.def.sortOrder - b.def.sortOrder)
      .forEach(item => {
        features.push(item.def.displayName);
      });
  }
  
  return features;
}

File: backend/internal/subscription/model/feature_registry.go
package subscriptionmodel

import (
	"fmt"
	"strings"
	subscriptionconstants "kyooar/internal/subscription/constants"
)

type FeatureType = subscriptionconstants.FeatureType

type FeatureDefinition struct {
	Key           string                 `json:"key"`
	Type          FeatureType            `json:"type"`
	DisplayName   string                 `json:"display_name"`
	Description   string                 `json:"description"`
	Unit          string                 `json:"unit,omitempty"`
	UnlimitedText string                 `json:"unlimited_text,omitempty"`
	Format        string                 `json:"format,omitempty"`
	Icon          string                 `json:"icon,omitempty"`
	Category      string                 `json:"category,omitempty"`
	SortOrder     int                    `json:"sort_order"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
}

var FeatureRegistry = map[string]FeatureDefinition{
	subscriptionconstants.LimitOrganizations: {
		Key:           subscriptionconstants.LimitOrganizations,
		Type:          subscriptionconstants.FeatureTypeLimit,
		DisplayName:   "Organizations",
		Description:   "Maximum number of organizations",
		Unit:          "organizations",
		UnlimitedText: "Unlimited organizations",
		Format:        "{value} organization(s)",
		Icon:          "store",
		Category:      "core",
		SortOrder:     1,
	},
	subscriptionconstants.LimitQRCodes: {
		Key:           subscriptionconstants.LimitQRCodes,
		Type:          subscriptionconstants.FeatureTypeLimit,
		DisplayName:   "QR Codes",
		Description:   "Total QR codes across all organizations",
		Unit:          "QR codes",
		UnlimitedText: "Unlimited QR codes",
		Format:        "{value} QR codes",
		Icon:          "qr-code",
		Category:      "core",
		SortOrder:     2,
	},
	subscriptionconstants.LimitFeedbacksPerMonth: {
		Key:           subscriptionconstants.LimitFeedbacksPerMonth,
		Type:          subscriptionconstants.FeatureTypeLimit,
		DisplayName:   "Monthly Feedbacks",
		Description:   "Maximum feedbacks per month",
		Unit:          "feedbacks/month",
		UnlimitedText: "Unlimited feedbacks",
		Format:        "{value} feedbacks/month",
		Icon:          "message-square",
		Category:      "core",
		SortOrder:     3,
	},
	subscriptionconstants.LimitTeamMembers: {
		Key:           subscriptionconstants.LimitTeamMembers,
		Type:          subscriptionconstants.FeatureTypeLimit,
		DisplayName:   "Team Members",
		Description:   "Maximum team members",
		Unit:          "members",
		UnlimitedText: "Unlimited team members",
		Format:        "{value} team member(s)",
		Icon:          "users",
		Category:      "collaboration",
		SortOrder:     4,
	},

	subscriptionconstants.FlagBasicAnalytics: {
		Key:         subscriptionconstants.FlagBasicAnalytics,
		Type:        subscriptionconstants.FeatureTypeFlag,
		DisplayName: "Basic Analytics",
		Description: "View feedback analytics and insights",
		Icon:        "bar-chart-2",
		Category:    "analytics",
		SortOrder:   20,
	},
	subscriptionconstants.FlagAdvancedAnalytics: {
		Key:         subscriptionconstants.FlagAdvancedAnalytics,
		Type:        subscriptionconstants.FeatureTypeFlag,
		DisplayName: "Advanced Analytics",
		Description: "Detailed insights and reporting",
		Icon:        "bar-chart",
		Category:    "analytics",
		SortOrder:   21,
	},
	subscriptionconstants.FlagFeedbackExplorer: {
		Key:         subscriptionconstants.FlagFeedbackExplorer,
		Type:        subscriptionconstants.FeatureTypeFlag,
		DisplayName: "Feedback Explorer",
		Description: "Browse and search all feedback",
		Icon:        "search",
		Category:    "analytics",
		SortOrder:   22,
	},
	subscriptionconstants.FlagCustomBranding: {
		Key:         subscriptionconstants.FlagCustomBranding,
		Type:        subscriptionconstants.FeatureTypeFlag,
		DisplayName: "Custom Branding",
		Description: "Customize with your brand",
		Icon:        "palette",
		Category:    "customization",
		SortOrder:   23,
	},
	subscriptionconstants.FlagPrioritySupport: {
		Key:         subscriptionconstants.FlagPrioritySupport,
		Type:        subscriptionconstants.FeatureTypeFlag,
		DisplayName: "Priority Support",
		Description: "24/7 priority customer support",
		Icon:        "headphones",
		Category:    "support",
		SortOrder:   24,
	},
}

func GetFeatureDefinition(key string) (FeatureDefinition, bool) {
	def, exists := FeatureRegistry[key]
	return def, exists
}

func GetFeaturesByCategory(category string) []FeatureDefinition {
	var features []FeatureDefinition
	for _, def := range FeatureRegistry {
		if def.Category == category {
			features = append(features, def)
		}
	}
	return features
}

func FormatFeatureValue(key string, value interface{}) string {
	def, exists := FeatureRegistry[key]
	if !exists {
		return ""
	}

	switch def.Type {
	case subscriptionconstants.FeatureTypeLimit:
		limitValue, ok := value.(int64)
		if !ok {
			if intVal, ok := value.(int); ok {
				limitValue = int64(intVal)
			} else {
				return ""
			}
		}
		
		if limitValue == -1 {
			return def.UnlimitedText
		}
		
		result := def.Format
		if result == "" {
			result = "{value} {unit}"
		}
		result = replaceValue(result, "{value}", limitValue)
		result = replaceValue(result, "{unit}", def.Unit)
		return result
		
	case subscriptionconstants.FeatureTypeFlag:
		if boolVal, ok := value.(bool); ok && boolVal {
			return def.DisplayName
		}
		return ""
		
	default:
		return ""
	}
}

func replaceValue(s string, old string, value interface{}) string {
	return strings.ReplaceAll(s, old, fmt.Sprintf("%v", value))
}